buildkite/bazel-bench/generate_report.py

```diff
@@ -0,0 +1,259 @@
+# You may obtain a copy of the License at
+#
+#    http:#www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Generates a daily HTML report for the projects.
+The steps:
+  1. Get the necessary data from Storage for projects/date.
+  2. Manipulate the data to a format suitable for graphs.
+  3. Generate a HTML report containing the graphs.
+  4. Upload the generated HTMLs to GCP Storage.
+"""
+import argparse
+import collections
+import csv
+import datetime
+import json
+import io
+import os
+import statistics
+import subprocess
+import sys
+import tempfile
+import urllib.request
+
+
+def _platform_path_str(posix_path):
+    """Converts the path to the appropriate format for platform."""
+    if os.name == "nt":
+        return posix_path.replace("/", "\\")
+    return posix_path
+
+
+# For the event markers in JSON profiles.
+# TODO(leba): Include JSON profiles data.
+EVENTS_ORDER = [
+    "Launch Blaze",
+    "Initialize command",
+    "Load packages",
+    "Analyze dependencies",
+    "Analyze licenses",
+    "Prepare for build",
+    "Build artifacts",
+    "Complete build",
+]
+TMP = tempfile.gettempdir()
+REPORTS_DIRECTORY = _platform_path_str("{}/.bazel_bench/reports".format(TMP))
+
+
+def _upload_to_storage(src_file_path, storage_bucket, destination_dir):
+    """Uploads the file from src_file_path to the specified location on Storage.
+  """
+    args = ["gsutil", "cp", src_file_path, "gs://{}/{}".format(storage_bucket, destination_dir)]
+    subprocess.run(args)
+
+
+def _load_csv_from_remote_file(http_url):
+    with urllib.request.urlopen(http_url) as resp:
+        reader = csv.DictReader(io.TextIOWrapper(resp))
+        return [row for row in reader]
+
+
+def _load_json_from_remote_file(http_url):
+    with urllib.request.urlopen(http_url) as resp:
+        data = resp.read()
+        encoding = resp.info().get_content_charset("utf-8")
+        return json.loads(data.decode(encoding))
+
+
+def _get_storage_url(storage_bucket, dated_subdir):
+    return "https://{}.storage.googleapis.com/{}".format(storage_bucket, dated_subdir)
+
+
+def _get_dated_subdir_for_project(project, date):
+    return "{}/{}".format(project, date.strftime("%Y/%m/%d"))
+
+
+def _prepare_data_for_graph(performance_data):
+    """Massage the data to fit a format suitable for graph generation.
+  """
+    ordered_commit_to_readings = collections.OrderedDict()
+    for entry in performance_data:
+        bazel_commit = entry["bazel_commit"]
+        if bazel_commit not in ordered_commit_to_readings:
+            ordered_commit_to_readings[bazel_commit] = {
+                "bazel_commit": bazel_commit,
```

Comment:
add a TODO somewhere where it makes sense to convert the commit text into links

Reply:
Done